File: dadaia_workspace/features/panel/views/sessions.py
"""Sessions dashboard sub-section view — aggregated-cost dashboard scaffold.

panel-plumbing v0.1.52 FR2 introduced the dashboard-only scaffold; v0.1.79
(panel agentic-layers reorg) relocated it from its own standalone primary tab
into a sub-section rendered INSIDE the "1º Agentic Layer" (``#section-subagents``)
tabpanel. ``#section-sessions`` survives as the nested mount id — ``sessions.js``
keys on it unchanged — but it is now a plain ``<div>`` sub-section (like
``.ops-subsection``), not a standalone ``role="tabpanel"``.

Security (OWASP A03):
  This module returns a static HTML scaffold only. No session data is serialised
  server-side into this HTML — the 4-card summary dashboard and the cost-unknown
  banner are rendered client-side by ``window.Sessions`` in ``sessions.js`` from
  the ``/api/sessions`` aggregate envelope, after the auth handshake. This
  prevents any accidental injection of API payloads into HTML.

FR2 (SPEC §FR2):
  The Sessions dashboard renders ONLY the 4-card summary dashboard
  (#sessions-dashboard) plus the cost-unknown banner (#sessions-banner). The list
  table, sortable headers, filter toolbar, and skeleton machinery were removed
  with the client-computed statistics — the aggregate now lives server-side and
  is served by ``/api/sessions`` (v0.1.52 FR1). The #sessions-last-updated badge,
  extracted from the deleted toolbar, reports the last aggregate refresh time.
"""

from __future__ import annotations


def render_sessions_section() -> str:
    """Return the static HTML scaffold for the dashboard-only Sessions sub-section.

    ``sessions.js`` fetches the ``/api/sessions`` aggregate on 1º Agentic Layer
    tab load and on every ``dadaia:runtime-change`` event, renders the four stat
    cards into #sessions-dashboard, toggles #sessions-banner for cost-unknown
    runtimes (codex/pi), and updates the #sessions-last-updated badge. No
    per-session rows are rendered.

    v0.1.79: rendered as a plain ``<div id="section-sessions" class="ops-subsection">``
    — NOT a standalone ``.section``/``.panel-section`` tabpanel — so it is shown or
    hidden by its PARENT tabpanel's (#section-subagents) ``.active`` toggle, never
    independently.
    """
    return (
        '<div id="section-sessions" class="ops-subsection">\n'
        '  <header class="section-header">\n'
        "    <h2>Sessions</h2>\n"
        '    <div class="runtime-switcher" role="radiogroup" aria-label="Active runtime">\n'
        '      <button type="button" class="runtime-btn runtime-btn--claude" id="sessions-runtime-btn-claude"\n'
        '        role="radio" aria-checked="true" data-runtime-value="claude" aria-label="Claude runtime">\n'
        '        <span class="runtime-btn-icon" aria-hidden="true">&#9672;</span>\n'
        '        <span class="runtime-btn-label">Claude</span>\n'
        "      </button>\n"
        '      <button type="button" class="runtime-btn runtime-btn--codex" id="sessions-runtime-btn-codex"\n'
        '        role="radio" aria-checked="false" data-runtime-value="codex" aria-label="Codex runtime">\n'
        '        <span class="runtime-btn-icon" aria-hidden="true">&#9667;</span>\n'
        '        <span class="runtime-btn-label">Codex</span>\n'
        "      </button>\n"
        '      <button type="button" class="runtime-btn runtime-btn--pi" id="sessions-runtime-btn-pi"\n'
        '        role="radio" aria-checked="false" data-runtime-value="pi" aria-label="PI runtime">\n'
        '        <span class="runtime-btn-icon" aria-hidden="true">&#9678;</span>\n'
        '        <span class="runtime-btn-label">PI</span>\n'
        "      </button>\n"
        "    </div>\n"
        "  </header>\n"
        # 4-card aggregate dashboard — populated by JS from the /api/sessions aggregate.
        '  <div id="sessions-dashboard" class="sessions-dashboard" aria-label="Sessions summary"></div>\n'
        # Last-updated badge — extracted from the deleted toolbar; JS refreshes it
        # after each aggregate fetch.
        '  <span id="sessions-last-updated" class="sessions-last-updated"\n'
        '        aria-live="polite" data-testid="sessions-last-updated">Never</span>\n'
        # Cost-unknown banner — hidden by default; JS shows it when the runtime is
        # codex or pi. aria-live="polite" so screen readers announce it when it appears.
        '  <div id="sessions-banner" class="sessions-banner"\n'
        '       role="status" aria-live="polite" hidden></div>\n'
        "</div>"
    )

File: dadaia_workspace/features/panel/views/assets/js/sessions.js
// sessions.js — Sessions tab UI (aggregated-cost dashboard only).
// panel-plumbing v0.1.52 FR2: the client list/table/drawer/sort/filter and the
// 10s auto-refresh were removed. The cost aggregate is computed server-side and
// served by GET /api/sessions; this module renders the 4-card summary + the
// cost-unknown banner from that envelope.
//
// Depends on: window.authedFetch() defined in core.js (loaded before this script)
//             window.Runtime (runtime.js) — get() + the dadaia:runtime-change event.
//
// API contract (SPEC §FR1 — normative):
//   GET /api/sessions?runtime=<runtime>
//     → {
//         runtime, total_sessions, active_sessions,
//         total_cost_usd (float | null), cost_known (bool),
//         total_messages, top_agent: {name, session_count} | null, generated_at
//       }
//
// Cost render mapping (SPEC §FR1 matrix, preserved client-side):
//   - cost-unknown runtime (codex/kimi)  → 'N/A'
//   - claude, total_cost_usd === null  → '—'
//   - otherwise                        → '$X.XX' (incl. '$0.00' for 0.0)
//
// Behaviour summary:
//   - On DOMContentLoaded: locate #section-sessions; if absent, exit (no-op).
//   - fetchAggregate() appends ?runtime= from Runtime.get() (fallback 'claude').
//   - renderDashboard(agg): renders the 4 stat cards from the aggregate envelope.
//   - updateBanner(): shows #sessions-banner for cost-unknown runtimes only.
//   - Re-fetch on the 'dadaia:runtime-change' CustomEvent (updateBanner first).
//   - Updates the #sessions-last-updated badge after each aggregate fetch.

(function () {
  'use strict';

  // ── Constants ─────────────────────────────────────────────────────────────────
  var DEFAULT_RUNTIME = 'claude';
  var CODEX_BANNER_TEXT = 'Cost not tracked for Codex';
  var KIMI_BANNER_TEXT = 'Cost not tracked for Kimi';

  // Runtimes with no per-event pricing — cost is unknown and never fabricated.
  // Codex and Kimi share this posture; the Cost card renders 'N/A' and the banner
  // appears for both.
  function isCostUnknownRuntime(runtime) {
    return runtime === 'codex' || runtime === 'kimi-code';
  }

  // ── Module state ─────────────────────────────────────────────────────────────
  var _loaded = false;         // has at least one successful aggregate fetch completed?

  // ── HTML escape helper ────────────────────────────────────────────────────────
  function escHtml(s) {
    return String(s == null ? '' : s).replace(/[&<>"]/g, function (c) {
      return { '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;' }[c];
    });
  }

  // ── Runtime resolution ────────────────────────────────────────────────────────
  function getRuntime() {
    if (window.Runtime && typeof window.Runtime.get === 'function') {
      return window.Runtime.get();
    }
    return DEFAULT_RUNTIME;
  }

  // ── Compact integer formatter (k / M) ─────────────────────────────────────────
  function fmtCompact(n) {
    n = Number(n) || 0;
    if (n >= 1000000) { return (n / 1000000).toFixed(1) + 'M'; }
    if (n >= 1000) { return (n / 1000).toFixed(1) + 'k'; }
    return String(n);
  }

  // ── Banner: show/hide the "Cost not tracked for …" notice ────────────────────
  // Called before each fetchAggregate() and on every dadaia:runtime-change. The
  // #sessions-banner element starts with the HTML [hidden] attribute (set in the
  // scaffold); JS shows it only when the runtime is cost-unknown (codex/kimi).
  function updateBanner() {
    var banner = document.getElementById('sessions-banner');
    if (!banner) { return; }
    var runtime = getRuntime();
    if (isCostUnknownRuntime(runtime)) {
      banner.textContent = (runtime === 'kimi-code') ? KIMI_BANNER_TEXT : CODEX_BANNER_TEXT;
      banner.removeAttribute('hidden');
    } else {
      banner.textContent = '';
      banner.setAttribute('hidden', '');
    }
  }

  // ── Dashboard: render the 4 stat cards from the server aggregate ─────────────
  // `agg` is the /api/sessions envelope, or null before the first fetch resolves
  // (in which case placeholder zeros/dashes are rendered).
  function renderDashboard(agg) {
    var dash = document.getElementById('sessions-dashboard');
    if (!dash) { return; }

    var runtime = getRuntime();
    var isCostUnknown = isCostUnknownRuntime(runtime);

    var totalSessions = agg ? agg.total_sessions : 0;
    var activeSessions = agg ? agg.active_sessions : 0;
    var totalCostUsd = agg ? agg.total_cost_usd : null;
    var totalMessages = agg ? agg.total_messages : 0;
    var topAgent = agg ? agg.top_agent : null;

    // Cost render mapping (SPEC §FR1): 'N/A' for cost-unknown runtimes, '—' when
    // claude cost is null, '$X.XX' otherwise (incl. '$0.00' for a known 0.0).
    var costVal = isCostUnknown ? 'N/A'
      : (totalCostUsd == null ? '—' : '$' + Number(totalCostUsd).toFixed(2));
    var costClass = (!isCostUnknown && totalCostUsd != null)
      ? 'sessions-stat-value sessions-stat-value--cost'
      : 'sessions-stat-value';
    var costSub = isCostUnknown
      ? ('not tracked for ' + (runtime === 'kimi-code' ? 'Kimi' : 'Codex'))
      : '&nbsp;';

    var activeSub = activeSessions > 0
      ? (escHtml(String(activeSessions)) + ' active')
      : 'none active';

    var turnsStr = fmtCompact(totalMessages);

    var agentName = (topAgent && topAgent.name) ? escHtml(topAgent.name) : '—';
    var agentCount = (topAgent && topAgent.session_count) ? topAgent.session_count : 0;
    var agentSub = (topAgent && agentCount > 0)
      ? escHtml(String(agentCount)) + ' session' + (agentCount !== 1 ? 's' : '')
      : '&nbsp;';

    dash.innerHTML =
      '<div class="sessions-stat-card">'
        + '<span class="sessions-stat-label">Total Sessions</span>'
        + '<span class="sessions-stat-value">' + escHtml(String(totalSessions)) + '</span>'
        + '<span class="sessions-stat-sub">' + activeSub + '</span>'
      + '</div>'
      + '<div class="sessions-stat-card">'
        + '<span class="sessions-stat-label">Total Cost</span>'
        + '<span class="' + costClass + '">' + escHtml(costVal) + '</span>'
        + '<span class="sessions-stat-sub">' + costSub + '</span>'
      + '</div>'
      + '<div class="sessions-stat-card">'
        + '<span class="sessions-stat-label">AI Turns</span>'
        + '<span class="sessions-stat-value">' + escHtml(turnsStr) + '</span>'
        + '<span class="sessions-stat-sub">&nbsp;</span>'
      + '</div>'
      + '<div class="sessions-stat-card">'
        + '<span class="sessions-stat-label">Top Agent</span>'
        + '<span class="sessions-stat-value" style="font-size:1.05rem;line-height:1.3;">' + agentName + '</span>'
        + '<span class="sessions-stat-sub">' + agentSub + '</span>'
      + '</div>';
  }

  // ── Update the "Last updated" badge ────────────────────────────────────────────
  function updateLastUpdatedBadge() {
    var badge = document.getElementById('sessions-last-updated');
    if (!badge) { return; }
    badge.textContent = 'Updated: ' + new Date().toLocaleTimeString();
  }

  // ── Fetch the server-side aggregate and render the dashboard ─────────────────
  function fetchAggregate() {
    var runtime = getRuntime();

    window.authedFetch('/api/sessions?runtime=' + encodeURIComponent(runtime))
      .then(function (r) {
        if (!r.ok) { return null; }
        return r.json();
      })
      .then(function (agg) {
        if (!agg) {
          renderDashboard(null);
          updateLastUpdatedBadge();
          return;
        }
        _loaded = true;
        updateBanner();
        renderDashboard(agg);
        updateLastUpdatedBadge();
      })
      .catch(function () {
        renderDashboard(null);
      });
  }

  // ── Initialise ────────────────────────────────────────────────────────────────
  document.addEventListener('DOMContentLoaded', function () {
    var section = document.getElementById('section-sessions');
    if (!section) { return; } // No sessions section in the DOM — no-op

    // Set banner state and a cleared dashboard before the first fetch resolves.
    updateBanner();
    renderDashboard(null);

    // Re-fetch when the global runtime changes (runtime.js dispatches the event).
    // updateBanner() runs first so the banner (or its absence) is visible before
    // the fetch round-trip completes.
    document.addEventListener('dadaia:runtime-change', function () {
      updateBanner();
      renderDashboard(null);
      fetchAggregate();
    });

    // Initial fetch.
    fetchAggregate();
  });

  // ── Public API ────────────────────────────────────────────────────────────────
  window.Sessions = {
    isLoaded: function () { return _loaded; },
    reload: function () { fetchAggregate(); },
  };

})();
